src/elements.ts

```diff
@@ -1354,51 +1354,105 @@ const SVG_TAGS = ['path', 'rect', 'circle', 'ellipse', 'polygon', 'polyline',
   'g', 'defs', 'marker', 'line', 'text', 'tspan', 'pattern', 'mask', 'svg',
   'foreignObject', 'image', 'use'];
 
+type SvgTag =
+  'path'|
+  'rect'|
+  'circle'|
+  'ellipse'|
+  'polygon'|
+  'polyline'|
+  'g'|
+  'defs'|
+  'marker'|
+  'line'|
+  'text'|
+  'tspan'|
+  'pattern'|
+  'mask'|
+  'svg'|
+  'foreignObject'|
+  'image'|
+  'use';
+
+type InputTag =
+  'input' |
+  'select' |
+  'textarea';
+
+type MediaTag = 'video' | 'audio';
+
+type DomName =
+  SvgTag |
+  InputTag |
+  MediaTag |
+  string;
+
+type DomQuery = Element | DomName;
+
+type CreateResult<T extends DomName> =
+  T extends 'svg' ? SVGParentView :
+  T extends SvgTag ? SVGView :
+  T extends 'canvas' ? CanvasView :
+  T extends 'form' ? FormView :
+  T extends InputTag ? InputView :
+  T extends MediaTag ? MediaView :
+  ElementView;
+
+type QueryResult<T extends DomQuery> =
+  T extends Element ? ElementView :
+  T extends string ? CreateResult<T> | undefined :
+  ElementView | undefined;
+
+type QueryResults<T extends DomQuery> =
+  T extends Element ? ElementView[] :
+  T extends string ? CreateResult<T>[] :
+  ElementView[];
+
 /**
  * Finds the Element that matches a specific CSS selector, or creates a new
  * Element wrapper around a native HTMLElement instance.
  */
-export function $(query?: string|Element,
-    context?: ElementView): ElementView|undefined {
-  if (!query) return undefined;
+export function $<T extends DomQuery>(query?: T,
+    context?: ElementView): QueryResult<T> {
+  if (!query) return undefined as QueryResult<T>;
```

Comment:
The type system won't correctly infer the return value as `QueryResult<T>`, how to address this is an ongoing discussion in the TypeScript development community. (For example, see: https://github.com/microsoft/TypeScript/issues/24929)